src/states/common/base.rs

```diff
@@ -279,12 +328,77 @@ pub trait Base: Display {
             }
         };
 
-        self.send_message_over_network(conn_info, message);
+        self.network_service_mut()
+            .targets_cache_mut()
+            .insert_message(msg_id, vec![conn_info.clone()], 1);
+
+        self.send_message_over_network(conn_info, &message, msg_id);
     }
 
-    fn send_message_over_network(&mut self, conn_info: ConnectionInfo, message: Message) {
+    fn send_message_to_targets(
+        &mut self,
+        dst_targets: &[PublicId],
+        dg_size: usize,
+        message: Message,
+    ) {
+        let next_id = self.network_service_mut().next_msg_id();
+        self.send_message_with_id_to_targets(dst_targets, dg_size, message, next_id);
+    }
+
+    fn send_message_with_id_to_targets(
+        &mut self,
+        dst_targets: &[PublicId],
+        dg_size: usize,
+        message: Message,
+        msg_id: u64,
+    ) {
+        let conn_infos: Vec<_> = dst_targets
+            .into_iter()
+            .filter_map(|pub_id| self.peer_map().get_connection_info(pub_id).cloned())
+            .collect();
+
+        // initially only send to dg_size targets
+        for conn_info in conn_infos.iter().take(dg_size) {
+            self.send_message_over_network(conn_info.clone(), &message, msg_id);
+        }
+
+        self.network_service_mut()
+            .targets_cache_mut()
+            .insert_message(msg_id, conn_infos, dg_size);
+    }
+
+    fn send_message_to_next_target(
+        &mut self,
+        msg: NetworkBytes,
+        msg_id: u64,
+        failed_tgt: SocketAddr,
+    ) {
+        if let Some(tgt) = self
```

Comment:
I'm wondering if the body of this function would not be better within network_service_mut
=>
```
self.network_service_mut().send_message_to_next_target(msg, msg_id, failed_tgt);
```

If it is for the log identifier, we have a patter with peer_manager where we give the identifier to the function so it can use it.